File: flutter_valorai/lib/features/auth/presentation/state/auth_session_manager.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../../data/datasources/auth_remote_data_source.dart';
import '../../domain/models/auth_session.dart';

enum AuthSessionStatus { uninitialized, anonymous, authenticated }

class AuthSessionManager extends ChangeNotifier {
  AuthSessionManager._({
    FirebaseAuth? firebaseAuth,
    FlutterSecureStorage? storage,
    AuthRemoteDataSource? remoteDataSource,
  }) : _firebaseAuth = firebaseAuth ?? FirebaseAuth.instance,
       _storage = storage ?? const FlutterSecureStorage(),
       _remoteDataSource = remoteDataSource ?? AuthRemoteDataSource();

  static final AuthSessionManager instance = AuthSessionManager._();

  static const _accessTokenKey = 'valorai_access_token';
  static const _expiresAtKey = 'valorai_access_token_expires_at';
  static const _userMetadataKey = 'valorai_user_metadata';

  final FirebaseAuth _firebaseAuth;
  final FlutterSecureStorage _storage;
  final AuthRemoteDataSource _remoteDataSource;

  AuthSessionStatus _status = AuthSessionStatus.uninitialized;
  String? _accessToken;
  DateTime? _expiresAt;
  AuthUserMetadata? _user;
  Future<String>? _renewal;

  AuthSessionStatus get status => _status;
  String? get accessToken => _accessToken;
  AuthUserMetadata? get user => _user;
  bool get isAuthenticated =>
      _status == AuthSessionStatus.authenticated &&
      _accessToken != null &&
      _expiresAt != null &&
      _expiresAt!.isAfter(DateTime.now());

  Future<void> restoreSession() async {
    try {
      final token = await _storage.read(key: _accessTokenKey);
      final expiresAtRaw = await _storage.read(key: _expiresAtKey);
      final userRaw = await _storage.read(key: _userMetadataKey);
      final expiresAt = expiresAtRaw == null
          ? null
          : DateTime.tryParse(expiresAtRaw);

      if (token != null &&
          token.isNotEmpty &&
          expiresAt != null &&
          expiresAt.isAfter(DateTime.now()) &&
          userRaw != null) {
        _setAuthenticated(
          token: token,
          expiresAt: expiresAt,
          user: AuthUserMetadata.fromJson(userRaw),
        );
        return;
      }

      if (_firebaseAuth.currentUser != null) {
        await renewSession();
        return;
      }
    } on Exception {
      await _clearStoredSession();
    }

    _setAnonymous();
  }

  Future<String> completeFirebaseSignIn() {
    return renewSession();
  }

  Future<String> renewSession() {
    final activeRenewal = _renewal;
    if (activeRenewal != null) {
      return activeRenewal;
    }

    final renewal = _exchangeFirebaseSession();
    _renewal = renewal;
    return renewal.whenComplete(() {
      _renewal = null;
    });
  }

  Future<String> _exchangeFirebaseSession() async {
    final firebaseUser = _firebaseAuth.currentUser;
    if (firebaseUser == null) {
      throw const AuthSessionException(
        'Sign in with Firebase before starting a ValorAI session.',
      );
    }

    final firebaseIdToken = await firebaseUser.getIdToken(true);
    if (firebaseIdToken == null || firebaseIdToken.isEmpty) {
      throw const AuthSessionException(
        'Firebase did not return an identity token.',
      );
    }

    final exchange = await _remoteDataSource.exchangeFirebaseToken(
      firebaseIdToken,
    );
    final expiresAt = DateTime.now().add(Duration(seconds: exchange.expiresIn));
    await _storage.write(key: _accessTokenKey, value: exchange.accessToken);
    await _storage.write(
      key: _expiresAtKey,
      value: expiresAt.toIso8601String(),
    );
    await _storage.write(key: _userMetadataKey, value: exchange.user.toJson());
    _setAuthenticated(
      token: exchange.accessToken,
      expiresAt: expiresAt,
      user: exchange.user,
    );
    return exchange.accessToken;
  }

  Future<void> expireSession() async {
    await _clearStoredSession();
    _setAnonymous();
  }

  Future<void> logout() async {
    await expireSession();
    await _firebaseAuth.signOut();
  }

  Future<void> _clearStoredSession() async {
    await Future.wait([
      _storage.delete(key: _accessTokenKey),
      _storage.delete(key: _expiresAtKey),
      _storage.delete(key: _userMetadataKey),
    ]);
    _accessToken = null;
    _expiresAt = null;
    _user = null;
  }

  void _setAuthenticated({
    required String token,
    required DateTime expiresAt,
    required AuthUserMetadata user,
  }) {
    _accessToken = token;
    _expiresAt = expiresAt;
    _user = user;
    _status = AuthSessionStatus.authenticated;
    notifyListeners();
  }

  void _setAnonymous() {
    _accessToken = null;
    _expiresAt = null;
    _user = null;
    _status = AuthSessionStatus.anonymous;
    notifyListeners();
  }
}

File: frontend/web/src/auth/authSessionManager.ts
import {
  getCurrentFirebaseUser,
  sendFirebasePasswordReset,
  signInFirebase,
  signOutFirebase,
  signUpFirebase,
  type FirebaseUserLike,
} from "./firebaseClient";
import { AuthSessionError, exchangeFirebaseToken, type TokenExchangeResult, type ValorUser } from "./tokenExchange";

export type AuthSessionStatus = "uninitialized" | "loading" | "anonymous" | "authenticated" | "expired";

export interface AuthSnapshot {
  status: AuthSessionStatus;
  accessToken: string | null;
  expiresAt: string | null;
  user: ValorUser | null;
  error: string | null;
  accessDenied: string | null;
}

interface StoredSession {
  accessToken: string;
  expiresAt: string;
  user: ValorUser;
}

interface StorageLike {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

interface AuthSessionDependencies {
  storage?: StorageLike;
  now?: () => number;
  exchangeFirebaseToken?: (firebaseIdToken: string) => Promise<TokenExchangeResult>;
  getCurrentFirebaseUser?: () => Promise<FirebaseUserLike | null>;
  signInFirebase?: (email: string, password: string) => Promise<FirebaseUserLike>;
  signUpFirebase?: (email: string, password: string, displayName?: string) => Promise<FirebaseUserLike>;
  sendPasswordReset?: (email: string) => Promise<void>;
  signOutFirebase?: () => Promise<void>;
}

const ACCESS_TOKEN_KEY = "valorai_access_token";
const EXPIRES_AT_KEY = "valorai_access_token_expires_at";
const USER_METADATA_KEY = "valorai_user_metadata";
const REFRESH_WINDOW_MS = 60_000;

function browserSessionStorage(): StorageLike | undefined {
  if (typeof window === "undefined") return undefined;
  return window.sessionStorage;
}

function safeParseSession(storage: StorageLike | undefined): StoredSession | null {
  if (!storage) return null;
  try {
    const accessToken = storage.getItem(ACCESS_TOKEN_KEY);
    const expiresAt = storage.getItem(EXPIRES_AT_KEY);
    const userRaw = storage.getItem(USER_METADATA_KEY);
    if (!accessToken || !expiresAt || !userRaw) return null;
    const user = JSON.parse(userRaw) as ValorUser;
    return { accessToken, expiresAt, user };
  } catch {
    return null;
  }
}

function messageForError(error: unknown): string {
  if (error instanceof AuthSessionError) return error.message;
  if (error instanceof Error && error.message.trim()) return error.message;
  return "ValorAI could not establish an authenticated session.";
}

function emptySnapshot(): AuthSnapshot {
  return {
    status: "uninitialized",
    accessToken: null,
    expiresAt: null,
    user: null,
    error: null,
    accessDenied: null,
  };
}

export class AuthSessionManager {
  private readonly storage?: StorageLike;
  private readonly now: () => number;
  private readonly exchangeToken: (firebaseIdToken: string) => Promise<TokenExchangeResult>;
  private readonly readFirebaseUser: () => Promise<FirebaseUserLike | null>;
  private readonly signInWithFirebase: (email: string, password: string) => Promise<FirebaseUserLike>;
  private readonly signUpWithFirebase: (email: string, password: string, displayName?: string) => Promise<FirebaseUserLike>;
  private readonly resetPasswordWithFirebase: (email: string) => Promise<void>;
  private readonly signOutFromFirebase: () => Promise<void>;
  private readonly listeners = new Set<(snapshot: AuthSnapshot) => void>();

  private snapshotState: AuthSnapshot = emptySnapshot();
  private bootstrapTask: Promise<void> | null = null;
  private renewalTask: Promise<string> | null = null;

  constructor(dependencies: AuthSessionDependencies = {}) {
    this.storage = dependencies.storage ?? browserSessionStorage();
    this.now = dependencies.now ?? (() => Date.now());
    this.exchangeToken = dependencies.exchangeFirebaseToken ?? exchangeFirebaseToken;
    this.readFirebaseUser = dependencies.getCurrentFirebaseUser ?? getCurrentFirebaseUser;
    this.signInWithFirebase = dependencies.signInFirebase ?? signInFirebase;
    this.signUpWithFirebase = dependencies.signUpFirebase ?? signUpFirebase;
    this.resetPasswordWithFirebase = dependencies.sendPasswordReset ?? sendFirebasePasswordReset;
    this.signOutFromFirebase = dependencies.signOutFirebase ?? signOutFirebase;
  }

  get snapshot(): AuthSnapshot {
    return this.snapshotState;
  }

  get accessToken(): string | null {
    return this.snapshotState.accessToken;
  }

  get isAuthenticated(): boolean {
    return this.hasUsableToken(this.snapshotState);
  }

  subscribe(listener: (snapshot: AuthSnapshot) => void): () => void {
    this.listeners.add(listener);
    listener(this.snapshotState);
    return () => {
      this.listeners.delete(listener);
    };
  }

  async restoreSession(): Promise<void> {
    if (this.bootstrapTask) {
      return this.bootstrapTask;
    }

    this.bootstrapTask = this.restoreSessionOnce().finally(() => {
      this.bootstrapTask = null;
    });
    return this.bootstrapTask;
  }

  async signIn(email: string, password: string): Promise<string> {
    this.setSnapshot({ ...this.snapshotState, status: "loading", error: null, accessDenied: null });
    try {
      await this.signInWithFirebase(email.trim(), password);
      return await this.renewSession();
    } catch (error) {
      this.clearStoredSession();
      this.setSnapshot({
        ...emptySnapshot(),
        status: "anonymous",
        error: messageForError(error),
      });
      throw error;
    }
  }

  async signUp(email: string, password: string, displayName?: string): Promise<string> {
    this.setSnapshot({ ...this.snapshotState, status: "loading", error: null, accessDenied: null });
    try {
      await this.signUpWithFirebase(email.trim(), password, displayName);
      return await this.renewSession();
    } catch (error) {
      this.clearStoredSession();
      this.setSnapshot({
        ...emptySnapshot(),
        status: "anonymous",
        error: messageForError(error),
      });
      throw error;
    }
  }

  async sendPasswordReset(email: string): Promise<void> {
    await this.resetPasswordWithFirebase(email.trim());
  }

  async getValidAccessToken(forceRefresh = false): Promise<string | null> {
    if (!forceRefresh && this.hasUsableToken(this.snapshotState)) {
      return this.snapshotState.accessToken;
    }

    const firebaseUser = await this.readFirebaseUser();
    if (!firebaseUser) {
      if (this.snapshotState.status === "authenticated") {
        await this.expireSession();
      }
      return null;
    }

    return this.renewSession();
  }

  async renewSession(): Promise<string> {
    if (this.renewalTask) return this.renewalTask;

    this.renewalTask = this.exchangeFirebaseSession().finally(() => {
      this.renewalTask = null;
    });

    return this.renewalTask;
  }

  async expireSession(message = "Your session has expired. Please sign in again."): Promise<void> {
    this.clearStoredSession();
    this.setSnapshot({
      ...emptySnapshot(),
      status: "expired",
      error: message,
    });
  }

  async logout(): Promise<void> {
    this.clearStoredSession();
    try {
      await this.signOutFromFirebase();
    } finally {
      this.setSnapshot({ ...emptySnapshot(), status: "anonymous" });
    }
  }

  markAccessDenied(message = "You do not have permission to perform this action."): void {
    this.setSnapshot({
      ...this.snapshotState,
      accessDenied: message,
    });
  }

  clearAccessDenied(): void {
    if (!this.snapshotState.accessDenied) return;
    this.setSnapshot({
      ...this.snapshotState,
      accessDenied: null,
    });
  }

  private async restoreSessionOnce(): Promise<void> {
    const storedSession = safeParseSession(this.storage);
    if (storedSession && this.hasUsableStoredSession(storedSession)) {
      this.setAuthenticated({
        accessToken: storedSession.accessToken,
        expiresAt: storedSession.expiresAt,
        user: storedSession.user,
      });
      return;
    }

    this.clearStoredSession();
    try {
      const firebaseUser = await this.readFirebaseUser();
      if (firebaseUser) {
        await this.renewSession();
        return;
      }
    } catch {
      this.clearStoredSession();
    }

    this.setSnapshot({ ...emptySnapshot(), status: "anonymous" });
  }

  private async exchangeFirebaseSession(): Promise<string> {
    const firebaseUser = await this.readFirebaseUser();
    if (!firebaseUser) {
      throw new AuthSessionError("Sign in with Firebase before starting a ValorAI session.");
    }

    const firebaseIdToken = await firebaseUser.getIdToken(true);
    if (!firebaseIdToken) {
      throw new AuthSessionError("Firebase did not return an identity token.");
    }

    const exchange = await this.exchangeToken(firebaseIdToken);
    const expiresAt = new Date(this.now() + exchange.expires_in * 1000).toISOString();
    this.storeSession(exchange.access_token, expiresAt, exchange.user);
    this.setAuthenticated({
      accessToken: exchange.access_token,
      expiresAt,
      user: exchange.user,
    });
    return exchange.access_token;
  }

  private hasUsableToken(snapshot: AuthSnapshot): snapshot is AuthSnapshot & { accessToken: string; expiresAt: string } {
    if (snapshot.status !== "authenticated" || !snapshot.accessToken || !snapshot.expiresAt) return false;
    const expiresAt = Date.parse(snapshot.expiresAt);
    return Number.isFinite(expiresAt) && expiresAt - this.now() > REFRESH_WINDOW_MS;
  }

  private hasUsableStoredSession(session: StoredSession): boolean {
    const expiresAt = Date.parse(session.expiresAt);
    return Boolean(session.accessToken) && Number.isFinite(expiresAt) && expiresAt - this.now() > REFRESH_WINDOW_MS;
  }

  private setAuthenticated(session: { accessToken: string; expiresAt: string; user: ValorUser }): void {
    this.setSnapshot({
      status: "authenticated",
      accessToken: session.accessToken,
      expiresAt: session.expiresAt,
      user: session.user,
      error: null,
      accessDenied: null,
    });
  }

  private storeSession(accessToken: string, expiresAt: string, user: ValorUser): void {
    try {
      this.storage?.setItem(ACCESS_TOKEN_KEY, accessToken);
      this.storage?.setItem(EXPIRES_AT_KEY, expiresAt);
      this.storage?.setItem(USER_METADATA_KEY, JSON.stringify(user));
    } catch {
      // Browser storage can be unavailable in private modes; keep the in-memory session.
    }
  }

  private clearStoredSession(): void {
    try {
      this.storage?.removeItem(ACCESS_TOKEN_KEY);
      this.storage?.removeItem(EXPIRES_AT_KEY);
      this.storage?.removeItem(USER_METADATA_KEY);
    } catch {
      // Ignored for environments that expose but reject storage access.
    }
  }

  private setSnapshot(snapshot: AuthSnapshot): void {
    this.snapshotState = snapshot;
    for (const listener of this.listeners) {
      listener(snapshot);
    }
  }
}

export const authSessionManager = new AuthSessionManager();
